vault/custom_response_headers_test.go

```diff
@@ -0,0 +1,179 @@
+	}
+
+	if !lch.ExistCustomResponseHeader("X-Custom-Header") {
+		t.Fatalf("header name with X-Vault prefix is not valid")
+	}
+
+}
+
+func TestCustomResponseHeadersConfigInteractUiConfig(t *testing.T) {
+	b := testSystemBackend(t)
+	_, barrier, _ := mockBarrier(t)
+	view := NewBarrierView(barrier, "")
+	b.(*SystemBackend).Core.systemBarrierView = view
+
+	logger := logging.NewVaultLogger(log.Trace)
+	rawListenerConfig :=  []*configutil.Listener {
+		{
+			Type:    "tcp",
+			Address: "127.0.0.1:443",
+			CustomResponseHeaders: map[string]map[string]string{
+				"default": defaultCustomHeaders,
+				"307": customHeaders307,
+				"3xx": customHeader3xx,
+				"200": customHeaders200,
+				"2xx": customHeader2xx,
+				"400": customHeader400,
+			},
+		},
+	}
+	uiHeaders, err := b.(*SystemBackend).Core.uiConfig.Headers(context.Background())
+	if err != nil {
+		t.Fatalf("failed to get headers from ui config")
+	}
+	customListenerHeader := NewListenerCustomHeader(rawListenerConfig, logger, uiHeaders)
+	if customListenerHeader == nil {
+		t.Fatalf("custom header config should be configured")
+	}
+	b.(*SystemBackend).Core.customListenerHeader = customListenerHeader
+	clh := b.(*SystemBackend).Core.customListenerHeader
+	if clh == nil {
+		t.Fatalf("custom header config should be configured in core")
+	}
+
+	w := httptest.NewRecorder()
+	hw := logical.NewHTTPResponseWriter(w)
+
+	// setting a header that already exist in custom headers
+	req := logical.TestRequest(t, logical.UpdateOperation, "config/ui/headers/X-Custom-Header")
+	req.Data["values"] = []string{"UI Custom Header"}
+	req.ResponseWriter = hw
+
+	resp, err := b.HandleRequest(namespace.RootContext(nil), req)
+	if err == nil {
+		t.Fatal("request did not fail on setting a header that is present in custom response headers")
+	}
+	if !strings.Contains(resp.Data["error"].(string), fmt.Sprintf("This header already exist in server configuration. %v", "X-Custom-Header")) {
+		t.Fatalf("failed to get the expected error")
+	}
+
+	// setting a header that already exist in custom headers
+	req = logical.TestRequest(t, logical.UpdateOperation, "config/ui/headers/Someheader-400")
+	req.Data["values"] = []string{"400"}
+	req.ResponseWriter = hw
+
+	_, err = b.HandleRequest(namespace.RootContext(nil), req)
+	if err == nil {
+		t.Fatal("request did not fail on setting a header that is present in custom response headers")
+	}
+	h, err := b.(*SystemBackend).Core.uiConfig.Headers(context.Background())
+	if h.Get("Someheader-400") == "400" {
+		t.Fatalf("should not be able to set a header that is in custom response headers")
+	}
+
+	req = logical.TestRequest(t, logical.UpdateOperation, "config/ui/headers/X-CustomUiHeader")
+	req.Data["values"] = []string{"Ui header value"}
+	req.ResponseWriter = hw
+
+	_, err = b.HandleRequest(namespace.RootContext(nil), req)
+	if err != nil {
+		t.Fatal("request did not fail on setting a header that is present in custom response headers")
```

Comment:
You say "err != nil", but the message is "did not fail" - something seems wrong there.